### PR TITLE
[FSDP] Support FSDP with EP

### PR DESCRIPTION
**Pull request by Xing-lil**
### PR Category

Distributed Strategy

### PR Types

New features

### Description

Support FSDP with EP

### 是否引起精度变化

否

### CODE REVIEW

File: python/paddle/distributed/fsdp/fully_shard_fusion.py [hunk 1]
```diff
@@ -447,9 +447,10 @@ def all_gather_params(self, params, is_backward=False):
             tmp_buffer = params_buffer.get_tmp_buffer()
             # Do all_gather in sync: FREED -> USING
             if params_buffer.status == BufferState.FREED:
-                fsdp_group.process_group.all_gather(
-                    params_buffer.data_buffer, tmp_buffer
-                ).wait()
+                if fsdp_group.nranks > 1:
```
Comment (risemeup1111) [inline python/paddle/distributed/fsdp/fully_shard_fusion.py]: P1 `nranks == 1` 时这里跳过了唯一一次把参数值写入 `tmp_buffer` 的操作。构造阶段只在 `data_buffer` 中保留完整参数，而 `get_tmp_buffer()` 新建的是全零 Tensor；后面又把真实参数绑定到该临时缓冲。因此本 PR 新增的 `moe_sharding_degree=1` 专家组在首个 forward 就会使用全零权重。

请在单 rank 分支把 `data_buffer` 复制或共享到 `tmp_buffer`（也可以直接复用 `data_buffer`），并增加首个 FSDP forward 与 Stage1 参数或输出一致性的断言。

File: python/paddle/distributed/fsdp/fully_shard_fusion.py [hunk 2]
```diff
@@ -548,10 +559,20 @@ def _wait_for_grad_comm(self, queue_limit=2):
 
     def finish_grads_sync(self):
         # Wait for all async reduce_scatter tasks, call before optimizer.step()
+        groups = self.buffer_manager.buffer_groups
+        for group in reversed(groups):
+            if group.is_expert_param and not group.grads_reduced:
```
Comment (risemeup1111) [inline python/paddle/distributed/fsdp/fully_shard_fusion.py]: P1 `grads_reduced` 只在 `optimizer.step()` 触发的 `reset_params_buffer_status()` 中清零，因而会跨多次 backward 持久化。做梯度累积时，第一次 backward 已归约的 expert 组会保持为 `True`；第二次 backward 虽然在 grad hook 中用 `copy_` 写入了新梯度，但结构化 flush 和这里都会跳过该组，新的梯度不会进入 fused grad buffer。最早一个 FSDP unit 的临时梯度还会在下一次 backward 被直接覆盖。

请把 expert 组的归约状态限定在单次 backward 内，保证每次 backward 完成后都 flush 本轮临时梯度，并补充连续两次 `loss.backward()` 后再 `optimizer.step()`、与 Stage1 对齐的回归用例。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
| 序号 | 位置 | 优先级 | 状态 |
|---:|---|---|:---:|
| 1 | [`fully_shard_fusion.py:450`](https://github.com/PaddlePaddle/Paddle/pull/79640#discussion_r3749862424) | P1 | ✅ |
| 2 | [`fully_shard_fusion.py:564`](https://github.com/PaddlePaddle/Paddle/pull/79640#discussion_r3749864847) | P1 | 🚧 |
| 3 | [PR 标题](https://github.com/PaddlePaddle/Paddle/pull/79640) | P3 | 🚧 |
| 4 | [`dygraph_group_fsdp.py:405`](https://github.com/PaddlePaddle/Paddle/blob/ae4fcbe65c665033e31dc12faf3fbc14c972686d/test/collective/fleet/dygraph_group_fsdp.py#L405) | P1 | 🚧 |

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: python/paddle/distributed/fsdp/fully_shard_fusion.py (hunk 1 above)
Comment (risemeup1111) [reply python/paddle/distributed/fsdp/fully_shard_fusion.py]: 已核对当前 head：`is_sharded` 与 `get_tmp_buffer()` 在 `nranks == 1` 时直接复用 `data_buffer`，FREED 分支也不再发起通信，因此该问题已修复。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

File: test/collective/fleet/test_dygraph_fsdp_for_eager.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    def test_dygraph_group_fsdp(self):
+        self.run_mnist_2accelerators('dygraph_group_fsdp.py')
+
+    # check dygraph fsdp + ep for some functions.
+    def test_dygraph_group_fsdp_moe(self):
+        self.run_mnist_2accelerators('dygraph_group_fsdp_moe.py')
+
+    # check dygraph fsdp + ep with expert params sharded inside the ep group.
+    def test_dygraph_group_fsdp_moe_sharding(self):
+        assert (
```
Comment (risemeup1111) [inline test/collective/fleet/test_dygraph_fsdp_for_eager.py]: P1 新增 target 在 `LOCAL_ALL_ARCH AND LOCAL_ALL_PLAT` 下注册，`testslist.csv` 该条目的 arch/os 为空且资源列仍是 2；因此 CPU/XPU、少于 4 张 GPU 或普通 2 卡调度都会执行这里，普通 `assert` 会直接使测试失败（同类 `run_mnist_2accelerators` 在设备不足时会跳过）。请将该用例限定为 GPU/4 卡的专用调度，或在资源不足时使用 `self.skipTest(...)`，避免非目标环境误报失败。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

File: python/paddle/distributed/fsdp/fully_shard_fusion.py (hunk 2 above)
Comment (risemeup1111) [reply python/paddle/distributed/fsdp/fully_shard_fusion.py]: 本次提交新增的 overlap/sync 对照只位于 `run_dense()`；`run_moe(ep_degree)` 仍未接入任何注册测试，且 `train_moe()` 仍在每次 `loss.backward()` 后立即 `optimizer.step()`。因此新增覆盖没有执行专家分片 reduce-scatter，也没有验证连续 backward 累积，本 P1 仍未解决。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

File: python/paddle/distributed/fsdp/fully_shard_fusion.py (hunk 2 above)
Comment (risemeup1111) [reply python/paddle/distributed/fsdp/fully_shard_fusion.py]: 本次提交仅合并 `develop`；`fully_shard_fusion.py`、`dygraph_group_fsdp.py` 及测试注册相关文件与合并前的 blob 完全一致。`run_moe(ep_degree)` 仍未接入注册测试，`train_moe()` 仍在每次 backward 后立即 step，因此本 P1 状态不变。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

File: python/paddle/distributed/fsdp/fully_shard_fusion.py (hunk 2 above)
Comment (risemeup1111) [reply python/paddle/distributed/fsdp/fully_shard_fusion.py]: 本次提交仍仅合并 `develop`；FSDP 实现、MoE 测试脚本及测试注册文件与上一 head 的 blob 完全一致。`run_moe(ep_degree)` 仍未接入注册测试，`train_moe()` 仍未覆盖连续 backward 后再 step，因此本 P1 状态不变。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

**Review by liym27 — APPROVED** [review]
LGTM

**Review by From00 — APPROVED** [review]
LGTM
